src/steamship/agents/agents.py

```diff
@@ -0,0 +1,187 @@
+        url = self.client._url(operation="package/instance/invoke")
+        payload = {
+            "instanceHandle": self._handle,
+            "payload": {
+                "httpVerb": "POST",
+                "invocationPath": self._method,
+                "arguments": task_kwargs,
+            },
+        }
+        return self._call(url, payload, time)
+
+    def _call(self, url, payload, schedule_time) -> Task:
+        # does the steamship-appropriate calling function.
+        pass
+
+
+class ExampleCacheTool(WorkspaceTool):
+    name = "CacheTool"
+    human_description = "In-Memory Key-Value Store"
+    ai_description = ("Used to store and lookup values by a known key.",)
+    store: KeyValueStore
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.store = KeyValueStore(client=self.client, store_identifier=f"cache-{uuid.uuid4()}")
+
+    def run(self, tool_input: List[Block], context: AgentContext) -> List[Block]:
+        pass
+
+
+class Agent(BaseModel, ABC):
+    tools: Optional[Tool] = []
+
+    @abstractmethod
+    def next_action(self, context: AgentContext) -> Action:
+        pass
+
+    @abstractmethod
+    def execute_action(self, action: Action, context: AgentContext):
+        pass
+
+    @abstractmethod
+    def run(self, agent_input: List[Block], context: AgentContext) -> List[Block]:
+        pass
+
+
+REACT_PROMPT = "<insert prompt that works here>."
+
+
+class ReACTAgent(Agent):
+    client: Steamship
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self._llm = self.client.use_plugin("gpt-4")
+
+    def _extract_action(self, blocks: [Block]) -> Action:
+        # whatever format matches the prompt...
+        pass
+
+    def next_action(self, context: AgentContext) -> Action:
+        # assuming most recent block is stored in context ??
+        action_task = self._llm.generate(...)
+        action_task.wait()
+        return self._extract_action(action_task.output.blocks)
+
+    def execute_action(self, action: Action, context: AgentContext):
+        tool_name = action.name
+        tool = self.tools.get(tool_name)
+        # todo: how to track new blocks? in context? somewhere else?
+        new_blocks = tool.run(action.data, context)
+        context.update_blocks(new_blocks)
+
+    def run(self, agent_input: List[Block], context: AgentContext) -> List[Block]:
+        # push agent_input somewhere?  into context?
+        while not isinstance(action := self.next_action(context), FinishAction):
+            self.execute_action(action=action, context=context)
+
+
+class ChatHistory(BaseModel):
+    pass
+
+
+class ChatReACTAgent(ReACTAgent):
+    client: Steamship
+    chat_history: ChatHistory
+
+    def __init__(self, chat_id: str, **kwargs):
+        super().__init__(**kwargs)
+        self.chat_history = File.create(client=self.client, handle=f"history-{chat_id}")
+
+
+class ExampleService(PackageService):
```

Comment:
I think I'm pretty sold on the idea that Agent owned by Package but not Package. And this general flavor of how they interact feels pretty good to me. 

Some high level thinking out loud:

- If agents want to schedule future work, they'll need the RequestContext and some notion of "invoke_later". That has the potential to turn into a weird encapsulation leak... but I not so concerned that it feels like a reason to make Agent = Package. 
- I think some standard notion of both input and output might make a ton of sense at the package-level. I'm not sure if it's `conversational_input` or `agent_input` or what -- but just to homogenize the comms to the point of plug-and-play telegram / debugging / agent wiring / Block marshalling, etc.

.. basically this direction here feels super comfortable to me.